java-checks/src/main/java/org/sonar/java/checks/security/CryptographicKeySizeCheck.java

```diff
@@ -0,0 +1,135 @@
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S4426")
+public class CryptographicKeySizeCheck extends AbstractMethodDetection {
+
+  private static final String KEY_PAIR_GENERATOR = "java.security.KeyPairGenerator";
+  private static final String KEY_GENERATOR = "javax.crypto.KeyGenerator";
+  private static final String GET_INSTANCE_METHOD = "getInstance";
+  private static final String STRING = "java.lang.String";
+
+  private static GetInstanceMethodInfo getInstanceInfo = new GetInstanceMethodInfo();
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition(KEY_GENERATOR).name(GET_INSTANCE_METHOD).addParameter(STRING),
+      MethodMatcher.create().typeDefinition(KEY_PAIR_GENERATOR).name(GET_INSTANCE_METHOD).addParameter(STRING),
+      MethodMatcher.create().typeDefinition(KEY_PAIR_GENERATOR).name("initialize").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(KEY_GENERATOR).name("init").withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (isGetInstanceMethod(mit)) {
+      storeSymbolOfAssignedInstance(mit);
+    } else if (isValidKeyGenSymbolOfInitializeMethod(mit)) {
+      checkAlgorithmParameterToReport(mit.arguments().get(0)).ifPresent(reportString -> reportIssue(mit, reportString));
+    }
+  }
+
+  private static boolean isValidKeyGenSymbolOfInitializeMethod(MethodInvocationTree mit) {
+    boolean isValidKeyGenSymbol = false;
+    if (mit.methodSelect().is(Tree.Kind.MEMBER_SELECT) && getInstanceInfo.keyGenMethodSymbol != null) {
+      ExpressionTree expr = ((MemberSelectExpressionTree) mit.methodSelect()).expression();
+      if (expr.is(Tree.Kind.IDENTIFIER)) {
+        isValidKeyGenSymbol = getInstanceInfo.keyGenMethodSymbol.equals(((IdentifierTree) expr).symbol());
+      } else if (expr.is(Tree.Kind.MEMBER_SELECT)) {
+        isValidKeyGenSymbol = getInstanceInfo.keyGenMethodSymbol.equals(((MemberSelectExpressionTree) expr).identifier().symbol());
+      }
+    }
+    return isValidKeyGenSymbol;
+  }
+
+  private static boolean isGetInstanceMethod(MethodInvocationTree mit) {
```

Comment:
it's not ideal to have side effects in methods, even if private.
This function basically does 2 things:

- checks if it's a get instance method of the known algorithms
- sets the global minKey size

Based on the [Single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle), this method does two much.

I suggest that this method should just return the algorithm type and a pass that type to `checkAlgorithmParameterToReport`. Then the `checkAlgorithmParameterToReport` method can use a dictionary declared top-level that maps the algorithm type to the minKey size value. Doing this you can get rid of the global state inside `GetInstanceMethodInfo.minKeySizeValue`

Reply:
I agree with the single responsibility principle and that a dictionary could show clearer the relationship between the algorithm and the key size.

But regarding the algorithm, when ` getInstance ` function is found, I need to store the name of the algorithm(here I store the minKeySize) and the `Symbol` of the variable that the ` KeyPairGenetaror ` / ` KeyGenerator ` is assigned to, because the issue is raised when an `init`/`initialize` function is invoked.
When, `init`/`initialize` is invoked I need to check if the `Symbol` of the variable that calls the function is the same as the one stored and then according to the algorithm that I have found from `geInstance` and the argument of `init`/`initialize` raise an issue or not.
IMO I prefer having the Object `GetInstanceMethodInfo`, than having variables defined globally.